### PR TITLE
New: Various HealthCheck Improvements

### PR DESCRIPTION
**Pull request by bakerboy448**
#### Database Migration
NO

#### Description

- New: RemotePath HealthCheck
- misc healthcheck improvements to sync downstream with Radarr
- HealthCheckFixtureExtensions add wikiFragment check to ShouldBeError

#### Todos
- [x] Tests
- [ ] Wiki Updates

#### Issues Fixed or Closed by this PR

*

### CODE REVIEW

File: src/NzbDrone.Core/HealthCheck/Checks/ImportListRootFolderCheck.cs [hunk 1]
```diff
@@ -40,7 +41,7 @@ public override HealthCheck Check()
                     continue;
                 }
 
-                if (!_diskProvider.FolderExists(rootFolderPath))
+                if (rootFolderPath.IsNullOrWhiteSpace() || !_diskProvider.FolderExists(rootFolderPath))
```
Comment (markus101) [inline src/NzbDrone.Core/HealthCheck/Checks/ImportListRootFolderCheck.cs:44]: Is there a real world case where the root folder path is null or white space?

File: src/NzbDrone.Core/HealthCheck/Checks/RemotePathMappingCheck.cs [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+                    if (dlpath.IsNullOrWhiteSpace())
+                    {
+                        return new HealthCheck(GetType(), HealthCheckResult.Error, "Sonarr failed to import (an) episode(s).  Check your logs for details.", "#remote-path-import-failed");
+                    }
+
+                    if (!dlpath.IsPathValid())
+                    {
+                        if (!status.IsLocalhost)
+                        {
+                            return new HealthCheck(GetType(), HealthCheckResult.Error, string.Format("Remote download client {0} reported files in {1} but this is not a valid {2} path.  Review your remote path mappings and download client settings.", client.Definition.Name, dlpath, _osInfo.Name), "#bad-remote-path-mapping");
```
Comment (markus101) [inline src/NzbDrone.Core/HealthCheck/Checks/RemotePathMappingCheck.cs:169]: What's with all the double spaces between sentences? Please update to a single space

File: src/NzbDrone.Core/HealthCheck/Checks/RemotePathMappingCheck.cs [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+                {
+                    var status = client.GetStatus();
+                    var folders = status.OutputRootFolders;
+                    foreach (var folder in folders)
+                    {
+                        if (!folder.IsValid)
+                        {
+                            if (!status.IsLocalhost)
+                            {
+                                return new HealthCheck(GetType(), HealthCheckResult.Error, string.Format("Remote download client {0} places downloads in {1} but this is not a valid {2} path.  Review your remote path mappings and download client settings.", client.Definition.Name, folder.FullPath, _osInfo.Name), "#bad-remote-path-mapping");
```
Comment (markus101) [inline src/NzbDrone.Core/HealthCheck/Checks/RemotePathMappingCheck.cs:69]: Same thing with double spaces here.

**Review by markus101 — CHANGES_REQUESTED** [review]

File: src/NzbDrone.Core/HealthCheck/Checks/RemotePathMappingCheck.cs (hunk 3 above)
Comment (markus101) [reply src/NzbDrone.Core/HealthCheck/Checks/RemotePathMappingCheck.cs:69]: Same issue here still (and others below).

Comment (bakerboy448) [reply src/NzbDrone.Core/HealthCheck/Checks/RemotePathMappingCheck.cs]: 3rd times the charm; sb good now

File: src/NzbDrone.Core/HealthCheck/Checks/DownloadClientRootFolderCheck.cs [hunk 4]
```diff
@@ -55,9 +57,13 @@ public override HealthCheck Check()
...
                     _logger.Debug(ex, "Unable to communicate with {0}", client.Definition.Name);
                 }
+                catch (HttpRequestException ex)
+                {
+                    _logger.Debug(ex, "Unable to communicate with {0}", client.Definition.Name);
+                }
                 catch (Exception ex)
                 {
-                    _logger.Error(ex, "Unknown error occurred in DownloadClientRootFolderCheck HealthCheck");
+                    _logger.Error(ex, "Unknown error occured in DownloadClientRootFolderCheck HealthCheck");
```
Comment (markus101) [inline src/NzbDrone.Core/HealthCheck/Checks/DownloadClientRootFolderCheck.cs:66]: image

occurred is correct according to google and various links from searching `occured`.

File: src/NzbDrone.Core/HealthCheck/Checks/DownloadClientRootFolderCheck.cs [hunk 5]
```diff
@@ -33,7 +34,9 @@ public class DownloadClientRootFolderCheck : HealthCheckBase, IProvideHealthChec
 
         public override HealthCheck Check()
         {
+            // Only check clients not in failure status, those get another message
             var clients = _downloadClientProvider.GetDownloadClients();
+
```
Comment (markus101) [inline src/NzbDrone.Core/HealthCheck/Checks/DownloadClientRootFolderCheck.cs:39]: No need to have an empty line between variable definitions.

File: src/NzbDrone.Core/HealthCheck/Checks/RemotePathMappingCheck.cs [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+            // Only check clients not in failure status, those get another message
+            var clients = _downloadClientProvider.GetDownloadClients();
+
+            foreach (var client in clients)
+            {
+                try
+                {
+                    var status = client.GetStatus();
+                    var folders = status.OutputRootFolders;
+                    foreach (var folder in folders)
```
Comment (markus101) [inline src/NzbDrone.Core/HealthCheck/Checks/RemotePathMappingCheck.cs:63]: new line before `foreach` block

File: src/NzbDrone.Core/HealthCheck/Checks/RemotePathMappingCheck.cs [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+
+            if (typeof(EpisodeImportFailedEvent).IsAssignableFrom(message.GetType()))
+            {
+                var failureMessage = (EpisodeImportFailedEvent)message;
+
+                // if we can see the file exists but the import failed then likely a permissions issue
+                if (failureMessage.EpisodeInfo != null)
+                {
+                    var episodePath = failureMessage.EpisodeInfo.Path;
+                    if (_diskProvider.FileExists(episodePath))
```
Comment (markus101) [inline src/NzbDrone.Core/HealthCheck/Checks/RemotePathMappingCheck.cs:131]: Newline before `if` block.

**Review by markus101 — CHANGES_REQUESTED** [review]
A few smaller issues and please rebase on `develop` for the recent code analysis changes.

**Comment by markus101** [comment]
Thanks for the fixups

**Pull request merged**